feat(common): http client request metadata for use in interceptors

FDIM opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Currently there is no way to configure interceptors in http client on per request basis and people (including myself) are abusing headers to disable / customize the behavior of certain interceptor (e.g. logging / authentication). 

Issue Number: Fixes #18155


## What is the new behavior?
This PR introduces a `context` option to any request that allows you to supply arbitrary data in a type safe way:
```
export const CacheInterceptorOptions = new HttpContextToken<{cache: boolean}>({ cache: false });
// compiles
this.http.get('/path', { context: new HttpContext().set(CacheInterceptorOptions, { cache: true }) })
this.http.get('/path', { context: new HttpContext().set(CacheInterceptorOptions, { cache: false }) })

// error
this.http.get('/path', { context: new HttpContext().set(CacheInterceptorOptions, { 
  foo: 123 // Argument of type '{ foo: number; }' is not assignable to parameter of type '{ cache: boolean; }'.
})}) 
```
There are also `get`, `delete`, and `keys` methods on `HttpContext`. `set` and `remove` can be chained to simplify usage.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

~~There is another PR open since last year but have seen no activity since January - probably due to other comments on the issue and lost type safety. ~~

~~Also, `HttpInterceptorMetadata` class is very generic (except for the name), maybe something like that could belong in core (or common?) package if there is a need?~~

EDIT: Description has been updated to match latest implementation

FDIM commented:
I am still alive and interested in updating this PR, so let me know I should do anything

un33k commented:
@FDIM You may want to resolve the conflict if you are still around & interested.

FDIM commented:
PR is ready again, failure in circle ci seems to be unrelated to my change

agroves333 commented:
Hey guys. Where are we at with this? Any idea on why CI is failing?

un33k commented:
I asked the Angular team in ngConf 2019 about this feature. The answer was, that this was one of the most requested feature and they have to wait for things to settle before adding this. So, it might be very unlikely for them to accepting this PR as-is.

BobDankert commented:
@un33k - what does "wait for things to settle" refer to? Just curious as I'm not quite sure what that means.

un33k commented:
@BobDankert  go to this link and timeline 21:04 [ng-conf-panel](https://www.youtube.com/watch?v=PjHIqx0w4do)

DutchKevv commented:
Any update? 

Can I help somehow .. ?

CesarD commented:
Why is this PR not yet approved?!
Are you owners just stalling or wth?!

alxhub commented:
@un33k oh that was you at ng-conf, awesome!

To expand on my answer there a little bit:

One of the big design questions here is whether metadata should be mutable or immutable. There are pros and cons to both:

* immutable metadata avoids surprises when requests are retried via an Observable pipeline, where an interceptor will see metadata for a previously "failed" request. This can be especially confusing when a downstream interceptor added the metadata, but other aspects of the requests (e.g. the headers) do not reflect the same mutations.

* mutable metadata would allow an interceptor to do exactly this on purpose (an obvious example here is a counter of retries on a request)

My instinct there is that immutable is the way to go. It's how the rest of the API already works, and you can always add an object to the metadata early and make use of its interior mutability if needed.

So assuming that's the direction we decide to go, the question becomes how to implement it. We have two immutable "map" types in `HttpClient`: `HttpParams` and `HttpHeaders`. Each has custom behavior (lazy parsing, encoding, etc) but similar functionality around append/set/delete, as well as get. Currently these are separate implementations.

If we introduce a third immutable container for metadata, it makes sense to refactor `HttpParams` and `HttpHeaders` to share a common immutable "core" that could be used to build the new container too.

BobDankert commented:
Thank you @alxhub for the response. Immutable metadata would be sufficient for our projects use cases.

mlc-mlapis commented:
@[**alxhub**](https://github.com/alxhub)

> If we introduce a third immutable container for metadata, it makes sense to refactor `HttpParams` and `HttpHeaders` to share a common immutable "core" that could be used to build the new container too.

Yep, it seems like a logical and necessary step. But from the timeline point of view, do you expect some extra phases to do that? Can they significantly delay the realization?

un33k commented:
@alxhub Yep that was me asking the question and hopefully we can once again attend ng-config in person again soon. 

As far are immutability goes, I think it can be just like the HttpParams and/or HttpHeaders. 
The only diff would be to have a hidden interceptor to peel the data off so it never gets on the wire.

Right now, I piggyback some metadata onto HttpHeaders and have a no-op interceptor placed  at the end which is responsible to peel the data off.  It works well, but it would be great to have a native solution.

CesarD commented:
> @[**alxhub**](https://github.com/alxhub)
> 
> > If we introduce a third immutable container for metadata, it makes sense to refactor `HttpParams` and `HttpHeaders` to share a common immutable "core" that could be used to build the new container too.
> 
> Yep, it seems like a logical and necessary step. But from the timeline point of view, do you expect some extra phases to do that? Can they significantly delay the realization?

There's been already 2 years of this PR trying to go into release... Perhaps a quick and separate implementation to make it fully available for us and leave the refactor with all the nice details for later? Just for the sake of availability and readiness.

> Right now, I piggyback some metadata onto HttpHeaders

I do the exact same thing on my end. Have to add some extra header so I can use as flag for some functionality I need and then remove it so it doesn't get sent with the actual backend request.

david0603 commented:
bttt

un33k commented:
@FDIM please close this ticket, so I can open a new one. Otherwise the new one is going to be redirected to this one and we end up in the no-op loop.

FDIM commented:
@un33k I'm still interested to finish this, but I have no idea who can make the decision. So this is stuck until someone from core team gets back.

CodeTroopers commented:
@FDIM Thanks for your cool PR. I'm also interested by this feature.

Just a details, I will rename `meta` and `HttpInterceptorMetadata` by `context` and `HttpContext` since this object could be interpreted by something else than an interceptor. I think `HttpContext` is a term used for this in the  ASP.Net Core framework

alxhub commented:
@FDIM thanks for your continued work on this feature. I think there's still some work to be done on the design, but this is definitely functionality that's been missing from `HttpClient` for far too long. Let's iterate on this PR and fix that.

After reviewing the current version and considering the API at a high level, I think a very simple design is preferable. @CodeTroopers suggestion of the `HttpContext` terminology really helps frame the idea.

So here's what I'm thinking:

1) Every `HttpRequest`, when created, has an `HttpContext`, which is effectively a metadata map. If you don't provide one, an empty default context is created internally.

2) This context is mutable and stays with the request for as long as the request is live, even across multiple retries.

This second point is a departure from my previous intuition, which is that request metadata should use the same immutable API as the rest of `HttpRequest`. Thinking about the metadata as a "context" changed my mind here. The mental model would be that every request has one, and only one, context, which persists across the entire lifetime of the request, including any retry operations and any response. Through this context, the operation of interceptors can be coordinated, even across request retries.

The amount of code to implement this is surprisingly small. Here's my concept of what `HttpContext` might look like:

```typescript
export class HttpContextKey<T> {
    constructor(readonly defaultValue: T) {}
}

export class HttpContext {
    private map = new Map<HttpContextKey<unknown>, unknown>();

    read<K>(key: HttpContextKey<K>): K {
        return this.map.has(key) ? this.map.get(key) as K : key.defaultValue;
    }

    write<K>(key: HttpContextKey<K>, value: K): void {
        this.map.set(key, value);
    }

    keys(): IterableIterator<HttpContextKey<unknown>> {
        return this.map.keys();
    }
}
```

That's it. Each new `HttpRequest` would then be either initialized with a context or have one created for it, and each `HttpResponse` would carry through the context of its `HttpRequest`. Easy, and very little new code.

Note that using a `Map` keyed on the instances of `HttpContextKey` means that there's perfect type safety here - there's no chance for two keys to collide by using the same string id. Each key also carries a default value, which allows for unambiguous reads as well as enables convenient type inference when creating a key.

Beyond the implementation above, we'll need:

1) support for the context in the `HttpClient` testing framework
2) documentation and examples in the `HttpClient` docs

mlc-mlapis commented:
@alxhub I think that's good news for today. Everyone interested in this area will be pleased, and most importantly, it is a much easier way than you thought before, in terms of deep restructuring, etc. Thank you.

Lonli-Lokli commented:
@alxhub What about request.clone method, it should not create copy of request context but init with previous one, right?

FDIM commented:
@alxhub this feedback is very much appreciated and I completely agree with `context` idea - it is a perfect name for this use case.

I have now updated the code to use new terminology and simplified token and context classes. Implementation should match what you've suggested except for method names.

I still think that current proposed API with `set`, `get`, `delete` methods is more appropriate. Mainly because it matches the API of HttpHeaders closer and does not invent a new pattern. I hope that you'll agree with this approach.


Regarding 2 last points:
- I have looked at the code of `angular/common/http/testing` package, I can't really see which part should be improved to cover the context? Conceptually these values have no effect on actual http calls.
- When it comes to the docs, I've added `@usageNotes` for `HttpContext`. Of course it should probably be improved further, but before I do that I just want to know if that is all that is needed?

Thank you for taking the time to look a this

CodeTroopers commented:
@alxhub @FDIM I will go more far in the design to let a full control to developpers.
First by creating a base class for `HttpContext` and move into this class the `withCredentials` and `reportProgress` members of `HttpRequest`. Let the developper the possibility to override the `HttpContext`. Then create a HttpRequest as a generic class `HttpRequest<T extends HttpContext>`.
I am aware that the implementation is more complex but the developer will have more control and can extend the design

alxhub reviewed (changes requested):
@FDIM this looks really solid.

I'm debating whether `HttpResponse` should include the context as well. Right now I'm leaning towards the simplest implementation of the idea (which would mean no, we wouldn't add the context anywhere but `HttpRequest`).

@CodeTroopers That's an interesting idea, but I don't think we should go that far (at least at first) for several technical reasons:

* Changing the type signature of `HttpRequest` is a breaking change, which we want to avoid.
* `withCredentials` and `reportProgress` aren't just "metadata", they directly affect the request being made. They don't belong in a context object.
* Having a single type for `HttpContext`, even if it's extensible by the developer, is actually a limitation we don't want.

To elaborate on that last point, one desirable property of the current design is that different libraries might provide interceptors that rely on the `HttpContext` metadata, and use it in different ways. They can do this because the current `HttpContext` mechanism of creating a type-safe `HttpContextKey` doesn't rely on listing all possible fields of the context object in one place.

By contrast, using a user-provided interface for `HttpContext` instead would mean that the end application developer would have to enumerate the required context fields from each library that wants to use context values, even if those context values are purely implementation details.

I think the current system is actually far more "extensible" as a result, as you don't _need_ to define a custom context type up front in order to store arbitrary data in the context.

Any application developer that wants to have a fully custom context object can always use this `HttpContext` to store and retrieve it, too.

The only other thing I'd request a change on is the commit message. I think it could be expanded a bit to describe the feature being built and the motivation behind it. For example, something like:

```
feat(compiler-cli): introduce HttpContext request context                                               
                                                                                                        
A long-requested feature for HttpClient is the ability to store and retrieve                            
custom metadata for requests, especially in interceptors. This commit                                   
implements this functionality via a new context object for requests.                                    
                                                                                                        
Each outgoing HttpRequest now has an associated "context", an instance of                               
the HttpContext class. An HttpContext can be provided when making a request,                            
or if not then an empty context is created for the new request. This context                            
shares its lifecycle with the entire request, even across operations that                               
change the identity of the HttpRequest instance such as RxJS retries.                                   
                                                                                                        
The HttpContext functions as an expando. Users can create typed keys as                                 
instances of HttpContextKey, and read/write a value for the key from any                                
HttpContext object.                                                                                     
                                                                                                        
This commit implements the HttpContext functionality. A followup commit will                            
add angular.io documentation.
```

If you can make these changes and get the CI green, I think this will be ready to merge in time for the 11.2 RC next week.

FDIM commented on packages/common/http/src/request.ts:
```diff
@@ -363,6 +388,9 @@ export class HttpRequest<T> {
     let params = update.params || this.params;
 
+    // Pass on context if needed
+    const context = update.context || this.context;
```
Yeah, I'm aware and I love that operator! I just matched the code in this context and I'm not too fond of updating things that are not related. I've updated this place though

dev054 commented on packages/common/http/src/request.ts:
```diff
@@ -196,6 +209,10 @@ export class HttpRequest<T> {
         this.headers = options.headers;
       }
 
+      if (!!options.context) {
```
I know that the other conditions use this, but this coercion is totally unnecessary.

FDIM replied:
@alxhub What is your take on these kind of changes? I'm fine doing them it just does not yield much benefit. It could most likely be fixed in entire codebase with an automatic fixer in eslint one day

alxhub replied:
Personally, I prefer explicit `!== undefined` checks. There _is_ a performance benefit to those, however, in framework code it's often not worth the increase in code size.

I have no preference between `if (options.context)` and `if (!!options.context)`.

FDIM commented:
@CodeTroopers I agree with @alxhub, HttpContext is probably the most flexible class to hold information. Original implementation I did still relied on string keys, last version does not. There is no way for interceptors/libraries to clash. This is a big concern for in the original issue.

When it comes to having access to the context from `HttpResponse`, I could see this useful but it is kind of a stretch. From usage point of view with current API:
- It would be fairly trivial to keep the reference to the context that user has passed if some interceptor would choose to have some kind of output. The caller is constructing a new instance after all.
- There is little reason to use this inside interceptor, as you always have access to the request object
- Only problematic case could be if the response processing is somewhere "far" from the caller and the interceptor. I vote for wait and see?

Now the build looks greener:
- there are some issues with increased size but I'm not sure if I should bump the number as part of this PR? Seems like a potential conflict 
- some build is failing due to changed public API. I'll try to resolve this one, I'm having issues running suggested command locally (and actually building/running tests). Will investigate next evening.

v11.2 sounds exciting, I've got like 6 interceptors that I could cleanup at work!

AndrewKushnir commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+ * @publicApi
+ */
+export class HttpContext {
+  private map = new Map<HttpContextToken<unknown>, unknown>();
```
I believe we can use `readonly` here:
```suggestion
  private readonly map = new Map<HttpContextToken<unknown>, unknown>();
```

alxhub commented:
@FDIM 

Re: the size increase, I think this is generally expected since we're adding a new class to `@angular/common/http` which is always going to be included. I can think of ways to design the API to avoid this, but they involve compromising its ergonomics, and I think ~400 bytes uncompressed for consumers of `HttpClient` is a very reasonable increase for the functionality being gained here. So it's reasonable to increase the payload size expectation in the test.

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+   *
+   * @returns A reference to itself for easy chaining
+   */
+  delete<T extends HttpContextToken<any>>(token: T): HttpContext {
```
There is no need for the generic type parameter here

```suggestion
  delete(token: HttpContextToken<any>): HttpContext {
```

dev054 replied:
Also we could use `unknown` here.

FDIM replied:
Updated to `delete<T>(token: HttpContextToken<T>)` to be consistent with others, TS fallback to the value being unknown anyway.

dev054 replied:
@FDIM Any reason to keep this generic type parameter? For other methods it **is** surely necessary, but for `delete` is useless.

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+   * Delete the value associated with given token
+   *
+   * @param token The reference to an instance of HttpContextToken
+   * @param value The value
```
This param is not applicable here.

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+   *
+   * @param token The reference to an instance of HttpContextToken
+   *
+   * @returns The stored value or default if one is defined
```
A token always has a default defined, wouldn't it?

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  /**
+   * Retrieve the value associated with the given token
```
nit: these docblock sentences should end with periods.

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+ */
+
+/**
+ * A token used to manipulate and access values stored in HttpContext
```
Type references are usually backticked.

```suggestion
 * A token used to manipulate and access values stored in `HttpContext`.
```

There are some more occurrences like these. Also notice the period at the end of the sentence :)

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  /**
+   * Delete the value associated with given token
```
```suggestion
   * Delete the value associated with the given token.
```

JoostK commented on packages/common/http/test/context_spec.ts:
```diff
@@ -0,0 +1,45 @@
+const IS_ENABLED = new HttpContextToken<boolean>(false);
+const CACHE_OPTION = new HttpContextToken<{cache: boolean, expiresIn?: number}>({cache: false});
+
+{
```
I notice this is done in all common/http tests, but this extra block isn't really necessary.

JoostK commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,97 @@
+ * Http context stores arbitrary user defined values and ensures type safety without
+ * actually knowing the types.
+ *
+ * This context is mutable and is shared between cloned requests.
```
Should this mention that cloned requests may have a context of their own, if provided?

alxhub commented on packages/common/http/test/context_spec.ts:
```diff
@@ -0,0 +1,45 @@
+      expect([...context.keys()]).toEqual([CACHE_OPTION]);
+
+      context.delete(CACHE_OPTION);
+      expect(context.get(CACHE_OPTION)).toEqual({cache: false});
```
👍 I think this is a worthy improvement.

FDIM replied:
Agree, this is a nice improvement. Most likely this will guarantee a sensible default value. As specifying `undefined` looks like this: 
`const IS_INTERCEPTOR_C_ENABLED = new HttpContextToken<boolean|undefined>(() => undefined);`

CodeTroopers commented:
@alxhub I agree with your point of view. But i disagree on this part:

> _`withCredentials` and `reportProgress` aren't just "metadata", they directly affect the request being made. They don't belong in a context object._

`HttpContext` is not  a simple metadata bag, it hold the informations that define the `HttRequest` behavior. `HttRequest` should contain only informations sent to the server (headers, body, token, params...). `withCredentials` and `reportProgress` define the behavior of `HttRequest` but they are not sent to the server. So they should be include in `HttpContext`

FDIM commented:
@alxhub PR should be ready now. Finally local build worked, I had to fallback and use git bash to run bazel.

No idea what causes this, but I cannot build project locally via powershell/cmd. Getting this very helpful error message:
```
---8<---8<--- Start of log, file at D:/_bazel_domas/nch2lzpe/bazel-workers/worker-2-TypeScriptCompile.log ---8<---8<---
'""' is not recognized as an internal or external command,
operable program or batch file.
---8<---8<--- End of log ---8<---8<---

````

alxhub commented:
> @alxhub I agree with your point of view. But i disagree on this part:
> 
> > _`withCredentials` and `reportProgress` aren't just "metadata", they directly affect the request being made. They don't belong in a context object._
> 
> `HttpContext` is not a simple metadata bag, it hold the informations that define the `HttRequest` behavior. `HttRequest` should contain only informations sent to the server (headers, body, token, params...). `withCredentials` and `reportProgress` define the behavior of `HttRequest` but they are not sent to the server. So they should be include in `HttpContext`

`withCredentials` definitely affects the shape of the outgoing request - I don't think your argument should apply to it. The boolean flag itself is not sent to the server, but it does influence what is.

As for `reportProgress`, I do understand where you're coming from. However, I still think that having a division between "well-known options that influence how `HttpClient` functions" and "user-defined metadata that influences how user-defined interceptors function" is a good design choice. `HttpContext` represents the latter - a container for options that are completely foreign to Angular itself.

alxhub commented on packages/common/http/src/client.ts:
```diff
@@ -11,12 +11,14 @@ import {Observable, of} from 'rxjs';
 import {HttpEvent, HttpResponse} from './response';
 
 
+
```
nit: extra newline

alxhub commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,96 @@
+ * Http context stores arbitrary user defined values and ensures type safety without
+ * actually knowing the types. It is backed by a `Map` and guarantees that keys do not clash.
+ *
+ * This context is mutable and is shared between cloned requests unless explicitely specified.
```
```suggestion
 * This context is mutable and is shared between cloned requests unless explicitly specified.
```

alxhub commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,96 @@
+   * @returns The stored value or default if one is defined.
+   */
+  get<T>(token: HttpContextToken<T>): T {
+    return this.map.has(token) ? this.map.get(token) as T : token.defaultValue();
```
This gets a little tricky now, with `defaultValue()` as a function. It means that each `get()` operation will return a new default value.

Instead, we should check if the key is present in the map, and set it to the default value if not:

```typescript
if (!this.map.has(token)) {
  this.map.set(token, token.defaultValue());
}
return this.map.get(token)! as T;
```

We should also have a test that verifies reading the default value twice gives back an identical value, even if the `defaultValue()` function for a token returns a new object each time.

FDIM replied:
Updated. It is tricky either way with this behavior. Now:
- default value becomes mutable within the same context. Could be good or bad, most likely should be documented.
- call to keys() will include this token even if nobody actually stored it, but only accessed. Maybe default values should be stored in secondary map?

FDIM replied:
Ping @alxhub, any input on the above comment?

alxhub replied:
Hey @FDIM,

I think both such behaviors are reasonable:

* I would expect to be able to do `context.get(SOME_KEY).field = true`. Essentially, it's as if _all_ keys are part of every context, they're just initialized lazily.
* `keys()` is more of a convenience method for iterating over the contents of an `HttpContext`. I think it's reasonable for it to return a key that was only ever accessed.

alxhub replied:
To put it another way, it'd be strange to me if `context.get(SOME_KEY)` returned a value and that value was _not_ in the map. Then `context.get(SOME_KEY) !== context.get(SOME_KEY)`, which doesn't make sense.

FDIM replied:
@alxhub Sounds good to me and what is even better is that this behavior matches the implementation. Hopefully this is ready to be merged besides the conflict that I'll resolve later today.

alxhub reviewed (approved):
Reviewed-For: public-api

petebacondarwin commented on packages/common/http/src/context.ts:
````diff
@@ -0,0 +1,99 @@
+ *
+ * ### Usage Example
+ *
+ * ```typescript
+ * // inside cache.interceptors.ts
+ * export const IS_CACHE_ENABLED = new HttpContextToken<boolean>(() => false);
+ *
+ * export class CacheInterceptor implements HttpInterceptor {
+ *
+ *   intercept(req: HttpRequest<any>, delegate: HttpHandler): Observable<HttpEvent<any>> {
+ *     if (req.context.get(IS_CACHE_ENABLED) === true) {
+ *       return ...;
+ *     }
+ *     return delegate.handle(req);
+ *   }
+ * }
+ *
+ * // inside a service
+ *
+ * this.httpClient.get('/api/weather', {
+ *   context: new HttpContext().set(IS_CACHE_ENABLED, true)
+ * }).subscribe(...);
+ * ```
````
It would be very nice if you could use doc-regions in a real code example and link to it here using `<code-example>` tags.

FDIM replied:
I'm not aware how these comments end up in documentation. Do you have any pointers where I can check this?

petebacondarwin replied:
So you would create a mini app that uses `HttpClient` in this folder: https://github.com/angular/angular/tree/master/packages/examples/http

And then you can mark up sections of the code using `docregion` comments. For example:

https://github.com/angular/angular/blob/b50d88073e7642df7d7b932fe4d388fcea2990d6/packages/examples/forms/ts/formBuilder/form_builder_example.ts#L46-L60

which you can then refer to in the docs. For example:

https://github.com/angular/angular/blob/28a0bcb424e4d64df0238986e34ed11adeeea866/packages/forms/src/form_builder.ts#L125-L126

FDIM replied:
Wow, this quite a setup. Could we have a separate issue for this? Feel free to assign me. I can see that http client has no examples there or bazel build configuration. Maybe I could add a general example that covers existing functionality and context API. Would this bring any value compared to what already exists for http client package?

petebacondarwin commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,99 @@
+   *
+   * @param token The reference to an instance of `HttpContextToken`.
+   *
+   * @returns The stored value or default if one is defined.
```
Perhaps explain that the default come from `token.defaultValue()`?

petebacondarwin commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,99 @@
+  keys(): IterableIterator<HttpContextToken<unknown>> {
+    return this.map.keys();
+  }
+}
```
I'm surprised that the formatter did not add a newline here.

petebacondarwin commented on packages/common/http/src/request.ts:
```diff
@@ -213,6 +230,11 @@ export class HttpRequest<T> {
     }
 
+    // If no context have been passed in, construct a new HttpContext instance.
+    if (!this.context) {
+      this.context = new HttpContext();
+    }
```
Given that this block guarantees that `this` will have a defined value for `this.context` by the end of the constructor, do we need to mark the property declaration above with a `!`?

FDIM replied:
Updated, let's see if it builds. Most likely non-issue anymore

FDIM replied:
Nope. Build failed:
```
packages/common/http/src/request.ts:234:15 - error TS2565: Property 'context' is used before being assigned.

234     if (!this.context) {
```

FDIM replied:
Not to nitpick but the last example will not compile. This one would:
```
this.context = options?.context ?? new HttpContext();
```
In current shape it matches the coding style used since the beginning, I'm not entirely sure if it's good idea to start mixing them now.
I could have a follow up issue to simplify the class itself, as it could be quite a bit shorter by using optional chaining and null coalescing. Not much would change for runtime code though, so reward is not that great for the effort.

petebacondarwin replied:
Ah that explains why you had the two separated blocks of code!
As I said this is not blocking merging this PR.

```ts
this.context = options && options.context || new HttpContext();
```

petebacondarwin commented on packages/common/http/test/module_spec.ts:
```diff
@@ -19,6 +20,8 @@ import {TestBed} from '@angular/core/testing';
 import {Observable} from 'rxjs';
 import {map} from 'rxjs/operators';
 
+const IS_INTERCEPTOR_C_ENABLED = new HttpContextToken<boolean|undefined>(() => undefined);
```
Why not this?

```suggestion
const IS_INTERCEPTOR_C_ENABLED = new HttpContextToken<boolean>(() => false);
```

FDIM replied:
Just to show case what it takes to have no default value. I bet default will always be `false` in actual apps.

petebacondarwin commented on packages/common/http/src/context.ts:
```diff
@@ -0,0 +1,99 @@
+ * Http context stores arbitrary user defined values and ensures type safety without
+ * actually knowing the types. It is backed by a `Map` and guarantees that keys do not clash.
+ *
+ * This context is mutable and is shared between cloned requests unless explicitly specified.
```
I didn't see one so I might have just missed it. But I think it would be good to have a test that demonstrates the context being reused between cloned requests?

FDIM replied:
Good catch, done

petebacondarwin reviewed (approved):
A bunch of non-blocking suggestions.
LGTM for public-api:

Reviewed-for: public-api

FDIM committed:
feat(compiler-cli): introduce HttpContext request context

A long-requested feature for HttpClient is the ability to store and retrieve
custom metadata for requests, especially in interceptors. This commit
implements this functionality via a new context object for requests.

Each outgoing HttpRequest now has an associated "context", an instance of
the HttpContext class. An HttpContext can be provided when making a request,
or if not then an empty context is created for the new request. This context
shares its lifecycle with the entire request, even across operations that
change the identity of the HttpRequest instance such as RxJS retries.

The HttpContext functions as an expando. Users can create typed tokens as instances of HttpContextToken, and
read/write a value for the key from any HttpContext object.

This commit implements the HttpContext functionality. A followup commit will
add angular.io documentation.

mary-poppins commented:
You can preview 2da0028 at https://pr25751-2da0028.ngbuilds.io/.
You can preview a439a18 at https://pr25751-a439a18.ngbuilds.io/.
You can preview 9688390 at https://pr25751-9688390.ngbuilds.io/.
You can preview 59ae2ee at https://pr25751-59ae2ee.ngbuilds.io/.
You can preview 053e2f0 at https://pr25751-053e2f0.ngbuilds.io/.
You can preview e92d6b6 at https://pr25751-e92d6b6.ngbuilds.io/.
You can preview d7197ce at https://pr25751-d7197ce.ngbuilds.io/.
You can preview ee82ca6 at https://pr25751-ee82ca6.ngbuilds.io/.
You can preview 6f64edb at https://pr25751-6f64edb.ngbuilds.io/.
You can preview 2fcb6f9 at https://pr25751-2fcb6f9.ngbuilds.io/.
You can preview a62a1ef at https://pr25751-a62a1ef.ngbuilds.io/.
You can preview 6c5c8c6 at https://pr25751-6c5c8c6.ngbuilds.io/.
You can preview 4f23a73 at https://pr25751-4f23a73.ngbuilds.io/.
You can preview e675268 at https://pr25751-e675268.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

un33k commented on packages/common/http/src/request.ts:
```diff
@@ -203,6 +216,10 @@ export class HttpRequest<T> {
         this.headers = options.headers;
       }
 
+      if (!!options.context) {
```
We keep using `!!options.context` and `!options.context` interchangeably.  It'd be great to just stick with `!options.context` when checking the `truthy` condition and only use `!!options.context` when we  need to assign the outcome to a boolean.

FDIM replied:
Well, this is checking if explicit context is provided and matches code style in this file.

Sorry to say but changing this place to !options.context does not really make sense and would change the meaning.

un33k commented on packages/common/http/src/request.ts:
```diff
@@ -213,6 +230,11 @@ export class HttpRequest<T> {
     }
 
+    // If no context have been passed in, construct a new HttpContext instance.
+    if (!this.context) {
```
This is the correct version of `!this.context` and strict eslint won't object it. Great!

un33k commented on packages/common/http/src/request.ts:
```diff
@@ -389,6 +417,7 @@ export class HttpRequest<T> {
     return new HttpRequest(method, url, body, {
       params,
       headers,
+      context,
```
I have been looking for `the` unit test where it ensures the context (aka browser metadata) doesn't make it to the server.  The `context` cannot be put on the wire, as the server may not know anything about it.  With that said, it'd be great to have a `unit test` to ensure the `context` would never make it out of the angular app. 

With that said, I am unsure where such `unit test` needs to go!, with the `request.ts` or `client.ts`. 

Please advise!

FDIM replied:
I somewhat understand the reasoning but usually unit tests are meant to cover cases on what should happen and not when something should not happen. The possibilities are endless for the latter and should be used sparingly.

If this in fact would be a problem, it would only mean that existing tests for request/client/body and interaction with http backend/xhr/fetch are too loose and allow extra properties. This in my opinion should not be allowed and would indicate a bigger problem - I hope this is not the case.

un33k replied:
`context` here is that of the ngApp, it may include a lot of sensitive state information. Without a unit test, we are hoping that `someone` doesn't by error include `context` in the outgoing request, in the future. 

If that happens, nothing would break before angular ships, cause we didn't have a test for it. However, internal guts of ngApp may be exposed and that may have some `security` implications.

With that said, feel free to resolve this .. and march forward.

FDIM replied:
I understand the implications, but as you said as well I have no idea where to put such test as what I've seen everything is passed explicitly when working with headers/body.
Input is welcome as I'll have to touch this PR once more anyway to resolve the conflict.
Thanks for the comments!

jessicajaniuk reviewed (approved):
LGTM 🍪 

reviewed-for: size-tracking

mary-poppins commented:
You can preview 4cba236 at https://pr25751-4cba236.ngbuilds.io/.

AndrewKushnir commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,17 +12,27 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 3033,
+<<<<<<< HEAD
```
FYI, there are merge conflict artifacts in this file, which causes some CI jobs to fail.

// cc @FDIM @alxhub

mary-poppins commented:
You can preview 7ec7207 at https://pr25751-7ec7207.ngbuilds.io/.

jessicajaniuk commented:
Caretaker Note: This has been merged.

un33k commented:
@FDIM You finally did it! Thank you very much to your perseverance, and ultimately this contribution.

FDIM commented:
I'm so happy that this finally made it into the framework and looking forward to cleaning up misused headers!

jcimoch commented:
so many years but it's finally there!!!! 
thanks @FDIM

nasreddineskandrani commented:
great work! this is big :)
